Mathlib/Algebra/GroupWithZero/Action/Basic.lean

```diff
@@ -0,0 +1,186 @@
+/-
+Copyright (c) 2018 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes, Yury Kudryashov
+-/
+import Mathlib.Algebra.Group.Action.Basic
+import Mathlib.Algebra.Group.Action.Prod
+import Mathlib.Algebra.Group.Aut
+import Mathlib.Algebra.GroupWithZero.Action.Defs
+import Mathlib.Algebra.GroupWithZero.Prod
+import Mathlib.Algebra.SMulWithZero
+
+/-!
+# Definitions of group actions
+
+This file defines a hierarchy of group action type-classes on top of the previously defined
+notation classes `SMul` and its additive version `VAdd`:
+
+* `MulAction M α` and its additive version `AddAction G P` are typeclasses used for
+  actions of multiplicative and additive monoids and groups; they extend notation classes
+  `SMul` and `VAdd` that are defined in `Algebra.Group.Defs`;
+* `DistribMulAction M A` is a typeclass for an action of a multiplicative monoid on
+  an additive monoid such that `a • (b + c) = a • b + a • c` and `a • 0 = 0`.
+
+The hierarchy is extended further by `Module`, defined elsewhere.
+
+Also provided are typeclasses for faithful and transitive actions, and typeclasses regarding the
+interaction of different group actions,
+
+* `SMulCommClass M N α` and its additive version `VAddCommClass M N α`;
+* `IsScalarTower M N α` and its additive version `VAddAssocClass M N α`;
+* `IsCentralScalar M α` and its additive version `IsCentralVAdd M N α`.
+
+## Notation
+
+- `a • b` is used as notation for `SMul.smul a b`.
+- `a +ᵥ b` is used as notation for `VAdd.vadd a b`.
+
+## Implementation details
+
+This file should avoid depending on other parts of `GroupTheory`, to avoid import cycles.
+More sophisticated lemmas belong in `GroupTheory.GroupAction`.
+
+## Tags
+
+group action
+-/
+
+-- TODO:
+-- assert_not_exists Ring
```

Comment:
Should this disappear, or be replaced with a different `assert_not_exists`?

Reply:
A future PR will actually enforce it, so I think it's perfectly fine as it is now. It shows the intent